src/compiler/checker.ts

```diff
@@ -11904,6 +11910,16 @@ namespace ts {
             return true;
         }
 
+        function isAritySmaller(sourceType: Type, target: Expression) {
+            if (isFunctionExpressionOrArrowFunction(target) && isFunctionType(sourceType)) {
+                const sourceSignatures = getSignaturesOfType(sourceType, SignatureKind.Call);
+                const sourceLengths = sourceSignatures.map(sig => !sig.hasRestParameter ? sig.parameters.length : Number.MAX_VALUE);
+                return forEach(sourceLengths, len => len < target.parameters.length);

```

Comment:
The previous commit took optional parameters into account, but @RyanCavanaugh pointed out that optional parameters are considered to be present for checking function parameters. The change is in a single commit so it's easy to revert if you think that optionality should be taken into account, though.